src/inference/src/dev/itensor.hpp

```diff
@@ -0,0 +1,133 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <memory>
+
+#include "openvino/core/coordinate.hpp"
+#include "openvino/core/core_visibility.hpp"
+#include "openvino/core/shape.hpp"
+#include "openvino/core/type/element_type.hpp"
+#include "openvino/runtime/allocator.hpp"
+
+namespace InferenceEngine {
+
+class Blob;
+
+}
+
+namespace ov {
+
+class ITensor : public std::enable_shared_from_this<ITensor> {
+public:
+    /**
+     * @brief Set new shape for tensor, deallocate/allocate if new total size is bigger than previous one.
+     * @note Memory allocation may happen
+     * @param shape A new shape
+     */
+    virtual void set_shape(const ov::Shape& shape);
```

Comment:
Maybe change to pass by value and then move (or add support to set by rvalue). It helps when set shape from temporary value.